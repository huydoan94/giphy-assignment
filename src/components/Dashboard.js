import React from 'react';
import { compose } from 'redux';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import {
  shape, arrayOf, func, bool, string,
} from 'prop-types';
import { Container, Grid, CircularProgress } from '@material-ui/core';
import { withStyles } from '@material-ui/core/styles';
import { get, isEmpty } from 'lodash';

import PictureBox from './dashboard/PictureBox';
import { actions } from './Dashboard.actions';
import { getPictures, getIsFetching } from './Dashboard.selectors';

import styles from './Dashboard.styles';

export class Dashboard extends React.PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      currentOffset: 0,
    };
  }

  componentDidMount() {
    this.fetchGiphy(this.state.currentOffset, 20, false);
  }

  fetchGiphy = (offset, limit, isAppend) => {
    this.setState({ currentOffset: offset + limit });
    this.props.fetchGiphy(offset, limit, isAppend);
  }

  handleScroll = (event) => {
    const { scrollTop, clientHeight, scrollHeight } = event.target;
    const isScrollNearBottom = scrollTop + clientHeight + 200 >= scrollHeight;
    if (isScrollNearBottom && !this.props.isFetching) {
      this.fetchGiphy(this.state.currentOffset, 20, true);
    }
  }

  makePictureProp = picture => ({
    id: picture.id,
    title: isEmpty(picture.title) ? 'Untitled' : picture.title,
    fixedHeightUrl: get(picture, 'images.fixed_height.url'),
    originUrl: get(picture, 'images.original.url'),
    trendingDate: picture.trending_datetime,
    username: get(picture, 'user.username', 'Unknown User'),
    userAvatarUrl: get(picture, 'user.avatar_url'),
  })

  render() {
    const { pictures, isFetching, classes } = this.props;

    return (
      <Container maxWidth="xl" className={classes.container} onScroll={this.handleScroll}>
        <Grid container spacing={2}>
          {pictures.map(picture => (
            <Grid item xs={12} sm={6} md={4} lg={3} key={picture.id}>
              <PictureBox {...this.makePictureProp(picture)} />
            </Grid>
          ))}
          {isFetching && (
            <Grid item xs={12} className={classes.progressContainer}>
              <CircularProgress className={classes.progress} />
            </Grid>
          )}
        </Grid>
      </Container>
    );
  }
}

Dashboard.propTypes = {
  pictures: arrayOf(shape({
    title: string,
    trending_datetime: string,
    user: shape({
      username: string,
      userAvatarUrl: string,
    }),
    images: shape({
      fixed_height: shape({
        url: string,
      }),
    }),
  })).isRequired,
  isFetching: bool.isRequired,
  fetchGiphy: func.isRequired,
  classes: shape({
    container: string,
    progressContainer: string,
    progress: string,
  }).isRequired,
};

const mapStatesToProps = createStructuredSelector({
  pictures: getPictures,
  isFetching: getIsFetching,
});

const mapDispatchersToProps = dispatch => ({
  fetchGiphy: (offset, limit = 100, isAppend = false) => dispatch(actions.fetchGiphy.start(offset, limit, isAppend)),
});

export default compose(
  withStyles(styles),
  connect(mapStatesToProps, mapDispatchersToProps),
)(Dashboard);
