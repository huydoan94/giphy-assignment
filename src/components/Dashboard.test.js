import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';

import { Dashboard } from './Dashboard';

describe('<Dashboard />', () => {
  let props;
  let component;
  let componentInstance;

  const createComponent = () => {
    component = shallow(<Dashboard {...props} />);
    componentInstance = component.instance();
  };

  beforeEach(() => {
    props = {
      pictures: [],
      isFetching: false,
      fetchGiphy: jest.fn(),
      classes: {},
    };
    createComponent();
  });

  it('should match snapshot', () => {
    expect(toJSON(component)).toMatchSnapshot();
  });

  it('should match snapshot when loading', () => {
    component.setProps({ isFetching: true });
    expect(toJSON(component)).toMatchSnapshot();
  });

  it('componentDidMount should call fetchGiphy', () => {
    componentInstance.fetchGiphy = jest.fn();
    const { currentOffset } = component.state();

    componentInstance.componentDidMount();
    expect(componentInstance.fetchGiphy).toHaveBeenCalledTimes(1);
    expect(componentInstance.fetchGiphy).toHaveBeenCalledWith(currentOffset, 20, false);
  });

  it('fetchGiphy should call prop fetchGiphy and set currentOffset', () => {
    props.fetchGiphy.mockReset();
    componentInstance.fetchGiphy(20, 20, true);

    expect(props.fetchGiphy).toHaveBeenCalledTimes(1);
    expect(props.fetchGiphy).toHaveBeenCalledWith(20, 20, true);
    expect(component.state().currentOffset).toBe(20 + 20);
  });

  it('handleScroll should not call fetchGiphy', () => {
    componentInstance.fetchGiphy = jest.fn();

    let eventTarget = {
      scrollTop: 0, clientHeight: 0, scrollHeight: 1000,
    };
    componentInstance.handleScroll({ target: eventTarget });
    expect(componentInstance.fetchGiphy).not.toHaveBeenCalled();

    eventTarget = {
      scrollTop: 500, clientHeight: 500, scrollHeight: 1000,
    };
    component.setProps({ isFetching: true });
    componentInstance.handleScroll({ target: eventTarget });
    expect(componentInstance.fetchGiphy).not.toHaveBeenCalled();
  });

  it('handleScroll should call fetchGiphy', () => {
    componentInstance.fetchGiphy = jest.fn();

    const { currentOffset } = component.state();
    const eventTarget = {
      scrollTop: 500, clientHeight: 500, scrollHeight: 1000,
    };
    componentInstance.handleScroll({ target: eventTarget });
    expect(componentInstance.fetchGiphy).toHaveBeenCalledTimes(1);
    expect(componentInstance.fetchGiphy).toHaveBeenCalledWith(currentOffset, 20, true);
  });

  it('makePictureProp should return correct props', () => {
    const data = {
      id: 'id',
      title: 'title',
      trending_datetime: 'trending_datetime',
      user: {
        username: 'username',
        avatar_url: 'avatar_url',
      },
      images: {
        original: {
          url: 'original_url',
        },
        fixed_height: {
          url: 'fixed_height_url',
        },
      },
    };
    const expected = {
      id: 'id',
      title: 'title',
      fixedHeightUrl: 'fixed_height_url',
      originUrl: 'original_url',
      trendingDate: 'trending_datetime',
      username: 'username',
      userAvatarUrl: 'avatar_url',
    };

    expect(componentInstance.makePictureProp(data)).toEqual(expected);
  });
});
