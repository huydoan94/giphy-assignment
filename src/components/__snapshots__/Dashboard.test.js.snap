// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dashboard /> should match snapshot 1`] = `
<WithStyles(ForwardRef(Container))
  maxWidth="xl"
  onScroll={[Function]}
>
  <WithStyles(ForwardRef(Grid))
    container={true}
    spacing={2}
  />
</WithStyles(ForwardRef(Container))>
`;

exports[`<Dashboard /> should match snapshot when loading 1`] = `
<WithStyles(ForwardRef(Container))
  maxWidth="xl"
  onScroll={[Function]}
>
  <WithStyles(ForwardRef(Grid))
    container={true}
    spacing={2}
  >
    <WithStyles(ForwardRef(Grid))
      item={true}
      sm={12}
    >
      <WithStyles(ForwardRef(CircularProgress)) />
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
</WithStyles(ForwardRef(Container))>
`;
